File: src/main/services/ModSourceLinker.ts
import fs from 'fs';
import path from 'path';
import type { GamebananaFileDto, GamebananaSourceDto, GamebananaSubmissionDto, ModItemDto, ModManifestDto, ModMetadataDto, ModUpdateInfoDto } from '../../shared/types';
import { BALDI_COMMUNITY_CATEGORY_ID } from '../../shared/types';
import { loadMetadata, saveMetadata, loadModManifest } from './ManifestLoader';
import { getSubmission, searchMods, downloadMod } from './GamebananaService';
import { installModArchive } from './ModInstaller';
import { createTempDir, extractArchive } from './ModArchiveExtractor';
const GMP_METADATA_FOLDER = '.rbbpmm';
const GMP_FALLBACK_METADATA_FOLDER = '_rbbpmm';
const METADATA_FILE = '.metadata';
function normalizeName(n: string): string {
    return (n || '')
        .toLowerCase()
        .replace(/\.[a-z0-9]+$/i, '')
        .replace(/[\s_\-()\[\]{}.,+]+/g, '')
        .trim();
}
function metadataFolderFor(modRoot: string): string | null {
    for (const f of [GMP_METADATA_FOLDER, GMP_FALLBACK_METADATA_FOLDER]) {
        const p = path.join(modRoot, f);
        if (fs.existsSync(p))
            return p;
    }
    return null;
}
function canPersist(mod: ModItemDto): boolean {
    if (mod.loose)
        return false;
    const folder = metadataFolderFor(mod.installDir);
    if (folder)
        return true;
    try {
        fs.mkdirSync(path.join(mod.installDir, GMP_METADATA_FOLDER), { recursive: true });
        return true;
    }
    catch {
        return false;
    }
}
function persistSource(mod: ModItemDto, manifest: ModManifestDto, source: GamebananaSourceDto): void {
    if (!canPersist(mod))
        return;
    const meta = loadMetadata(mod.installDir, manifest);
    meta.gamebananaSource = source;
    meta.installationUrl = `https://gamebanana.com/mods/${source.submissionId}`;
    saveMetadata(mod.installDir, manifest, meta);
}
function persistArchiveName(modRoot: string, manifest: ModManifestDto, archiveName: string): void {
    const folder = metadataFolderFor(modRoot);
    if (!folder)
        return;
    const meta = loadMetadata(modRoot, manifest);
    meta.lastInstalledArchiveName = archiveName;
    saveMetadata(modRoot, manifest, meta);
}
export { persistArchiveName };
function readSource(modRoot: string, manifest: ModManifestDto): GamebananaSourceDto | undefined {
    return loadMetadata(modRoot, manifest).gamebananaSource;
}
export function findMetadataFile(modRoot: string): string | null {
    const folder = metadataFolderFor(modRoot);
    if (!folder)
        return null;
    const p = path.join(folder, METADATA_FILE);
    return fs.existsSync(p) ? p : null;
}
export { GMP_METADATA_FOLDER, GMP_FALLBACK_METADATA_FOLDER };
function expectedNamesFor(mod: ModItemDto, manifest: ModManifestDto): string[] {
    const names: string[] = [];
    const meta = loadMetadata(mod.installDir, manifest);
    if (meta.lastInstalledArchiveName)
        names.push(meta.lastInstalledArchiveName);
    names.push(manifest.name);
    if (mod.directoryName)
        names.push(mod.directoryName);
    return names;
}
export async function searchCandidates(name: string): Promise<GamebananaSubmissionDto[]> {
    const res = await searchMods(1, name, BALDI_COMMUNITY_CATEGORY_ID);
    return (res.items ?? []).filter((i) => i.hasFiles);
}
export function matchFile(submission: GamebananaSubmissionDto, expectedNames: string[]): GamebananaFileDto | undefined {
    const all = [...(submission.files ?? []), ...(submission.archivedFiles ?? [])];
    if (all.length === 0)
        return undefined;
    const expected = expectedNames.map(normalizeName).filter(Boolean);
    let exact: GamebananaFileDto | undefined;
    for (const f of all) {
        if (!f.fileName)
            continue;
        const fn = normalizeName(f.fileName);
        if (expected.includes(fn)) {
            if (!exact || (f.dateAdded ?? 0) > (exact.dateAdded ?? 0))
                exact = f;
        }
    }
    if (exact)
        return exact;
    return [...all].sort((a, b) => (b.dateAdded ?? 0) - (a.dateAdded ?? 0))[0];
}
export async function linkByLocalMod(mod: ModItemDto, manifest: ModManifestDto): Promise<{
    linked: boolean;
    submission?: GamebananaSubmissionDto;
    file?: GamebananaFileDto;
}> {
    const candidates = await searchCandidates(mod.name || manifest.name);
    for (const sub of candidates) {
        const file = matchFile(sub, expectedNamesFor(mod, manifest));
        if (!file)
            continue;
        const source: GamebananaSourceDto = {
            submissionId: sub.id,
            fileId: file.id,
            fileName: file.fileName,
            version: file.version,
            submissionName: sub.name,
            dateLinked: new Date().toISOString()
        };
        persistSource(mod, manifest, source);
        return { linked: true, submission: sub, file };
    }
    return { linked: false };
}
export async function linkKnownSubmission(mod: ModItemDto, manifest: ModManifestDto, submission: GamebananaSubmissionDto, file: GamebananaFileDto): Promise<void> {
    const source: GamebananaSourceDto = {
        submissionId: submission.id,
        fileId: file.id,
        fileName: file.fileName,
        version: file.version,
        submissionName: submission.name,
        dateLinked: new Date().toISOString()
    };
    persistSource(mod, manifest, source);
}
export async function checkForUpdate(mod: ModItemDto, manifest: ModManifestDto): Promise<ModUpdateInfoDto> {
    const source = readSource(mod.installDir, manifest);
    if (!source) {
        return {
            hasUpdate: false,
            submissionId: 0,
            fileId: 0,
            fileName: '',
            downloadUrl: ''
        };
    }
    const sub = await getSubmission(source.submissionId);
    const all = [...(sub.files ?? []), ...(sub.archivedFiles ?? [])];
    const current = all.find((f) => f.id === source.fileId) ?? all[0];
    if (!current) {
        return {
            hasUpdate: false,
            submissionId: sub.id,
            fileId: source.fileId,
            fileName: source.fileName,
            downloadUrl: ''
        };
    }
    const hasUpdate = current.id !== source.fileId || (current.version ?? '') !== (source.version ?? '');
    return {
        hasUpdate,
        submissionId: sub.id,
        fileId: current.id,
        fileName: current.fileName,
        version: current.version,
        publishedDate: current.dateAdded,
        downloadUrl: current.downloadUrl
    };
}
export interface UpdateOutcome {
    ok: boolean;
    error?: string;
}
export async function updateMod(gameRoot: string, mod: ModItemDto, manifest: ModManifestDto, onProgress: (p: {
    stage: string;
    percent?: number;
    message?: string;
}) => void, isCancelled?: () => boolean): Promise<UpdateOutcome> {
    const source = readSource(mod.installDir, manifest);
    if (!source)
        return { ok: false, error: 'No linked source' };
    const sub = await getSubmission(source.submissionId);
    const all = [...(sub.files ?? []), ...(sub.archivedFiles ?? [])];
    const file = all.find((f) => f.id === source.fileId) ?? all[0];
    if (!file || !file.downloadUrl)
        return { ok: false, error: 'No downloadable file' };
    onProgress({ stage: 'downloading', percent: 0, message: `Downloading ${file.fileName}` });
    const archivePath = await downloadMod(file.downloadUrl, (p) => {
        const percent = p.total ? Math.min(99, Math.round((p.received / p.total) * 100)) : undefined;
        onProgress({ stage: 'downloading', percent, message: `Downloading ${file.fileName}` });
    }, isCancelled);
    let exTemp: string | null = null;
    const dotOld = mod.installDir + '.old';
    try {
        exTemp = createTempDir(mod.installDir);
        onProgress({ stage: 'extracting', message: 'Extracting archive' });
        const extractRoot = await extractArchive(archivePath, exTemp);
        if (fs.existsSync(dotOld))
            fs.rmSync(dotOld, { recursive: true, force: true });
        fs.renameSync(mod.installDir, dotOld);
        try {
            const result = await installModArchive(gameRoot, archivePath, undefined, onProgress, isCancelled, extractRoot);
            fs.rmSync(dotOld, { recursive: true, force: true });
            const newSource: GamebananaSourceDto = {
                ...source,
                fileId: file.id,
                fileName: file.fileName,
                version: file.version,
                dateLinked: new Date().toISOString()
            };
            const newManifest = result.mod ? loadModManifest(result.mod.installDir) ?? manifest : manifest;
            persistSource(result.mod, newManifest, newSource);
            return { ok: true };
        }
        catch (err) {
            if (fs.existsSync(dotOld) && !fs.existsSync(mod.installDir)) {
                fs.renameSync(dotOld, mod.installDir);
            }
            return { ok: false, error: err instanceof Error ? err.message : String(err) };
        }
    }
    finally {
        if (exTemp) {
            try {
                fs.rmSync(exTemp, { recursive: true, force: true });
            }
            catch {
            }
        }
        try {
            fs.rmSync(path.dirname(archivePath), { recursive: true, force: true });
        }
        catch {
        }
    }
}
export function getModMetadata(modRoot: string, manifest: ModManifestDto): ModMetadataDto {
    return loadMetadata(modRoot, manifest);
}

File: RBBPMM.Wpf/RBBPMM.Core/ModSourceLinker.cs
using System.Text.Json;

namespace RBBPMM.Core;

/// <summary>
/// Ties a locally installed mod to its GameBanana submission: links a source, checks for a newer
/// file, and performs an in-place update with a <c>.old</c> safety net.
/// Ported from src/main/services/ModSourceLinker.ts.
/// </summary>
public sealed class ModSourceLinker(GamebananaService banana)
{
    private readonly GamebananaService _banana = banana;

    public static string NormalizeName(string n) =>
        System.Text.RegularExpressions.Regex
            .Replace(
                System.Text.RegularExpressions.Regex.Replace((n ?? "").ToLowerInvariant(), @"\.[a-z0-9]+$", ""),
                @"[\s_\-()\[\]{}.,+]+", "")
            .Trim();

    private static string? MetadataFolderFor(string modRoot)
    {
        foreach (var f in new[] { Constants.GmpMetadataFolder, Constants.GmpFallbackMetadataFolder })
        {
            var p = Path.Combine(modRoot, f);
            if (Directory.Exists(p))
                return p;
        }
        return null;
    }

    public static string? FindMetadataFile(string modRoot)
    {
        var folder = MetadataFolderFor(modRoot);
        if (folder == null)
            return null;
        var p = Path.Combine(folder, Constants.MetadataFile);
        return File.Exists(p) ? p : null;
    }

    private static bool CanPersist(ModItem mod)
    {
        if (mod.Loose)
            return false;
        if (MetadataFolderFor(mod.InstallDir) != null)
            return true;
        try
        {
            Directory.CreateDirectory(Path.Combine(mod.InstallDir, Constants.GmpMetadataFolder));
            return true;
        }
        catch
        {
            return false;
        }
    }

    private static void PersistSource(ModItem mod, GamebananaSource source)
    {
        if (!CanPersist(mod))
            return;
        var meta = ManifestLoader.LoadMetadata(mod.InstallDir);
        meta.GamebananaSource = source;
        meta.InstallationUrl = $"https://gamebanana.com/mods/{source.SubmissionId}";
        ManifestLoader.SaveMetadata(mod.InstallDir, meta);
    }

    public static void PersistArchiveName(string modRoot, string archiveName)
    {
        if (MetadataFolderFor(modRoot) == null)
            return;
        var meta = ManifestLoader.LoadMetadata(modRoot);
        meta.LastInstalledArchiveName = archiveName;
        ManifestLoader.SaveMetadata(modRoot, meta);
    }

    public static GamebananaSource? ReadSource(string modRoot) =>
        ManifestLoader.LoadMetadata(modRoot).GamebananaSource;

    public static ModMetadata GetModMetadata(string modRoot) => ManifestLoader.LoadMetadata(modRoot);

    private static List<string> ExpectedNamesFor(ModItem mod)
    {
        var names = new List<string>();
        var meta = ManifestLoader.LoadMetadata(mod.InstallDir);
        if (!string.IsNullOrEmpty(meta.LastInstalledArchiveName))
            names.Add(meta.LastInstalledArchiveName!);
        names.Add(mod.Name);
        if (!string.IsNullOrEmpty(mod.DirectoryName))
            names.Add(mod.DirectoryName);
        return names;
    }

    public async Task<List<GamebananaSubmission>> SearchCandidatesAsync(string name, CancellationToken ct = default)
    {
        var res = await _banana.SearchMods(1, name, CategoryIds.BaldiCommunity, ct);
        return res.Items.Where(i => i.HasFiles).ToList();
    }

    /// <summary>Prefers a file whose name matches a known archive name, else the newest file.</summary>
    public static GamebananaFile? MatchFile(GamebananaSubmission submission, IReadOnlyList<string> expectedNames)
    {
        var all = submission.Files.Concat(submission.ArchivedFiles).ToList();
        if (all.Count == 0)
            return null;

        var expected = expectedNames.Select(NormalizeName).Where(n => n.Length > 0).ToList();

        GamebananaFile? exact = null;
        foreach (var f in all)
        {
            if (string.IsNullOrEmpty(f.FileName))
                continue;
            if (!expected.Contains(NormalizeName(f.FileName)))
                continue;
            if (exact == null || (f.DateAdded ?? 0) > (exact.DateAdded ?? 0))
                exact = f;
        }

        return exact ?? all.OrderByDescending(f => f.DateAdded ?? 0).First();
    }

    public sealed record LinkResult(bool Linked, GamebananaSubmission? Submission = null, GamebananaFile? File = null);

    /// <summary>Fuzzy-matches a local mod against GameBanana search results and persists the link.</summary>
    public async Task<LinkResult> LinkByLocalModAsync(ModItem mod, CancellationToken ct = default)
    {
        var candidates = await SearchCandidatesAsync(string.IsNullOrEmpty(mod.Name) ? mod.Name : mod.Name, ct);
        foreach (var sub in candidates)
        {
            var file = MatchFile(sub, ExpectedNamesFor(mod));
            if (file == null)
                continue;

            PersistSource(mod, new GamebananaSource
            {
                SubmissionId = sub.Id,
                FileId = file.Id,
                FileName = file.FileName,
                Version = file.Version,
                SubmissionName = sub.Name,
                DateLinked = DateTime.UtcNow.ToString("o")
            });
            return new LinkResult(true, sub, file);
        }

        return new LinkResult(false);
    }

    public Task LinkKnownSubmissionAsync(ModItem mod, GamebananaSubmission submission, GamebananaFile file)
    {
        PersistSource(mod, new GamebananaSource
        {
            SubmissionId = submission.Id,
            FileId = file.Id,
            FileName = file.FileName,
            Version = file.Version,
            SubmissionName = submission.Name,
            DateLinked = DateTime.UtcNow.ToString("o")
        });
        return Task.CompletedTask;
    }

    public async Task<ModUpdateInfo> CheckForUpdateAsync(ModItem mod, CancellationToken ct = default)
    {
        var source = ReadSource(mod.InstallDir);
        if (source == null)
            return new ModUpdateInfo(false, 0, 0, "", "");

        var sub = await _banana.GetSubmission(source.SubmissionId, ct);
        var all = sub.Files.Concat(sub.ArchivedFiles).ToList();
        var current = all.FirstOrDefault(f => f.Id == source.FileId) ?? all.FirstOrDefault();

        if (current == null)
            return new ModUpdateInfo(false, sub.Id, source.FileId, source.FileName, "");

        var hasUpdate = current.Id != source.FileId || (current.Version ?? "") != (source.Version ?? "");
        return new ModUpdateInfo(hasUpdate, sub.Id, current.Id, current.FileName, current.DownloadUrl)
        {
            Version = current.Version,
            PublishedDate = current.DateAdded
        };
    }

    /// <summary>
    /// Downloads the linked file over the installed mod. The old folder is parked at
    /// <c>&lt;installDir&gt;.old</c> and restored if the install throws.
    /// </summary>
    public async Task<Result<bool>> UpdateModAsync(
        string gameRoot,
        ModItem mod,
        Action<InstallProgress>? onProgress = null,
        Func<bool>? isCancelled = null,
        CancellationToken ct = default)
    {
        var source = ReadSource(mod.InstallDir);
        if (source == null)
            return Result<bool>.Fail("No linked source");

        var sub = await _banana.GetSubmission(source.SubmissionId, ct);
        var all = sub.Files.Concat(sub.ArchivedFiles).ToList();
        var file = all.FirstOrDefault(f => f.Id == source.FileId) ?? all.FirstOrDefault();
        if (file == null || string.IsNullOrEmpty(file.DownloadUrl))
            return Result<bool>.Fail("No downloadable file");

        onProgress?.Invoke(new InstallProgress("downloading", 0, $"Downloading {file.FileName}"));

        var downloadDir = Path.Combine(Path.GetTempPath(), "rbbpmm-update-" + Guid.NewGuid().ToString("N")[..8]);
        Directory.CreateDirectory(downloadDir);

        var archivePath = await _banana.DownloadFileAsync(
            file.DownloadUrl,
            downloadDir,
            file.FileName,
            new Progress<(long Received, long? Total)>(p =>
            {
                int? percent = p.Total is > 0 ? (int)Math.Min(99, p.Received * 100 / p.Total.Value) : null;
                onProgress?.Invoke(new InstallProgress("downloading", percent, $"Downloading {file.FileName}"));
            }),
            isCancelled);

        string? exTemp = null;
        var dotOld = mod.InstallDir + ".old";

        try
        {
            exTemp = ModArchiveExtractor.CreateTempDir(gameRoot);
            onProgress?.Invoke(new InstallProgress("extracting", null, "Extracting archive"));
            var extractRoot = await ModArchiveExtractor.ExtractArchiveAsync(archivePath, exTemp);

            if (Directory.Exists(dotOld))
                Directory.Delete(dotOld, recursive: true);
            Directory.Move(mod.InstallDir, dotOld);

            try
            {
                var result = ModInstaller.InstallModArchive(gameRoot, archivePath, null, isCancelled, extractRoot);
                Directory.Delete(dotOld, recursive: true);

                var newManifest = ManifestLoader.LoadModManifest(result.InstallDir);
                PersistSource(
                    new ModItem { InstallDir = result.InstallDir, Name = mod.Name, DirectoryName = mod.DirectoryName, Loose = false },
                    new GamebananaSource
                    {
                        SubmissionId = source.SubmissionId,
                        FileId = file.Id,
                        FileName = file.FileName,
                        Version = file.Version,
                        SubmissionName = source.SubmissionName,
                        DateLinked = DateTime.UtcNow.ToString("o")
                    });

                _ = newManifest;
                onProgress?.Invoke(new InstallProgress("done", 100, "Update installed"));
                return Result<bool>.Success(true);
            }
            catch (Exception ex)
            {
                if (Directory.Exists(dotOld) && !Directory.Exists(mod.InstallDir))
                    Directory.Move(dotOld, mod.InstallDir);
                return Result<bool>.Fail(ex.Message);
            }
        }
        catch (Exception ex)
        {
            return Result<bool>.Fail(ex.Message);
        }
        finally
        {
            if (exTemp != null)
                ModArchiveExtractor.RemoveDirIfInside(gameRoot, exTemp);
            try
            {
                if (Directory.Exists(downloadDir))
                    Directory.Delete(downloadDir, recursive: true);
            }
            catch
            {
                // best effort
            }
        }
    }
}
